### DISCUSSION TITLE
Validation error for (73,0): Type of field Annots is invalid: HexaPDF::Dictionary (HexaPDF::Error)

### DISCUSSION THREAD

**Discussion by work-wajeeh** [gettalong/hexapdf#283] [Q&A]
I'm trying to run this [Hello World](https://hexapdf.gettalong.org/examples/hello_world.html) example on a pdf document. Opened the document, made changes and when writing changes to the document, I'm getting this error. Anyone can help me understand/get it fixed?

```
require 'hexapdf'

doc = HexaPDF::Document.open(path_to_doc)
canvas = doc.pages.add.canvas
canvas.font('Helvetica', size: 100)
canvas.text("Hello World!", at: [20, 400])
doc.write("hello_world.pdf", optimize: true)
```
The error I get is

`hexapdf-0.36.0/lib/hexapdf/document.rb:719 'block in write': Validation error for (73,0): Type of field Annots is invalid: HexaPDF::Dictionary (HexaPDF::Error)`

When I try opening this document (before making any chnage) in Adobe Reader , this error pops up on one of the pages

<img width="264" alt="Screenshot 2024-01-23 at 3 34 57 AM" src="https://github.com/gettalong/hexapdf/assets/109785489/c9792410-dc86-4aa3-b67d-f54de3548651">

**Comment by gettalong** [comment]
So the document fails to show in Adobe Reader *before* you change in with HexaPDF? That would be consistent with HexaPDF showing the validation error.

In your case the validation error says that the field `/Annots`, which is one the fields a page can have, is of the wrong type. If you look at the documentation for the `Page` class at https://hexapdf.gettalong.org/documentation/api/HexaPDF/Type/Page.html, you can see the field definitions. And `/Annots` needs to be an array (or HexaPDF's wrapper class `PDFArray`) and not a PDF dictionary (hash in Ruby).

You can also run `hexapdf info --check path_to_file` to check for validation problems of the file.

If you can provide the PDF in question, I could have a look at it and maybe make it valid.

**Comment by work-wajeeh** [comment]
@gettalong sorry for responding you late. I'm sharing the PDF in the email. Thanks

**Reply by gettalong** [reply]
Thanks for the file! The file is indeed invalid because it references an invalid object as the annotation array. I will have to think about whether to apply a general solution for this or a specific one for just that case.

**Reply by work-wajeeh** [reply]
@gettalong were you able to check this issue?

**Reply by gettalong** [reply]
As I wrote I still need to think about a solution for this.

**Comment by matpowel** [comment]
A note on this @gettalong, this is happening to us sometimes even with PDFs created by HexaPDF. For example with the following code it's rare but does happen:

```   
    new_doc = HexaPDF::Document.new
    document_pages.order(:page_number).each_with_index do |document_page, i|
      next if document_page.page_number < split_at

      # Add this page to the new document unless we're at or beyond doc_end
      page = doc.pages[split_at - 1]
      new_doc.pages << new_doc.import(page) unless doc_end && document_page.page_number >= doc_end
      # Drop this page out of the first document, it's being moved to the new one
      doc.pages.delete(page)
    end
    new_doc.write(path) # Throws validation error (Annots is invalid)
```

**Reply by gettalong** [reply]
Thanks for the information! The problem here is most probably the same as in the other comments, i.e. that the source file (`doc` in your code sample) already has an invalid /Annots entry. It just gets copied as-is into the new document (`new_doc`) where it leads to the error.

**Reply by matpowel** [reply]
Hmm that shouldn't be possible. Our code checks the validity of every incoming PDF by calling doc.write() before we get into the splitting phase. If it fails we push it out to a service which extracts and re-creates a new PDF from scratch using the images etc.

We're on an old version of HexaPDF (0.40.0) because we've been restricted by an OpenSSL dependency, is there any chance a newer version would help?

**Reply by gettalong** [reply]
Probably not as the sample file I received together with the code from the original post still results in the error.

**Reply by gettalong** [reply]
> Hmm that shouldn't be possible. 

I'm not sure how HexaPDF would make the /Annots invalid when just copying a page to a new document. It would be great if you could share a PDF where this happens.

**Comment by gettalong** [comment]
@work-wajeeh Sorry for dropping the ball on this, it wasn't on my radar as it is a discussion and not an issue.

@work-wajeeh @matpowel It is hard to do this well in a general way as sometimes the type of a field might be invalid due to newer PDF versions or because the PDF uses deprecated things. I could add code especially for the /Annots field of a page but this problem applies to all fields of all PDF dictionary types.

Having said that, it is often the case that a field with a value of an invalid type is not used by most PDF viewers/readers. So removing the invalid value is probably the right way in many cases. I will add a new configuration option that would just do that but needs to be turned on manually, i.e. not by default.

**Reply by matpowel** [reply]
I think that's really sensible @gettalong, perhaps we can pass it to the validator. Is it possible to just have a mapping of default values for each of the non-critical properties and have an option to just set to default if an invalid one is found?

**Reply by gettalong** [reply]
If a field has a default value and HexaPDF knows about the specific dictionary type with that field, the default value is already defined. See e.g. https://github.com/gettalong/hexapdf/blob/master/lib/hexapdf/type/info.rb#L60-L61

So removing a set value would automatically mean falling back to the default value.

**Reply by gettalong** [reply]
FYI I implemented this now and it will be in the upcoming release, see https://github.com/gettalong/hexapdf/commit/6154bb486543babca5ed0ab59e56510308d8e05b on the devel branch. It is currently implemented in a non-configurable way as I think this will suffice.
